src/java/org/apache/cassandra/index/sai/disk/v1/TermsReader.java

```diff
@@ -203,6 +214,93 @@ public PostingsReader getPostingReader(long offset) throws IOException
         }
     }
 
+    public class RangeQuery
+    {
+        private final QueryEventListener.TrieIndexEventListener listener;
+        private final long lookupStartTime;
+        private final QueryContext context;
+
+        private final Expression exp;
+
+        RangeQuery(Expression exp, QueryEventListener.TrieIndexEventListener listener, QueryContext context)
+        {
+            this.listener = listener;
+            this.exp = exp;
+            lookupStartTime = System.nanoTime();
+            this.context = context;
+        }
+
+        public PostingList execute()
+        {
+            final ByteComparable lower = exp.lower != null ? ByteComparable.fixedLength(exp.getLowerBound()) : null;
+            final ByteComparable upper = exp.upper != null ? ByteComparable.fixedLength(exp.getUpperBound()) : null;
+            try (TrieRangeIterator reader = new TrieRangeIterator(termDictionaryFile.instantiateRebufferer(),
+                                                                  termDictionaryRoot,
+                                                                  lower,
+                                                                  upper,
+                                                                  exp.lowerInclusive,
+                                                                  exp.upperInclusive))
+            {
+                var iter = reader.iterator();
+                if (!iter.hasNext())
+                    return PostingList.EMPTY;
+
+                context.checkpoint();
+                // Because postings are not sorted, we need to eagerly materialize the results and sort them.
+                LongHeap postings = materializeResults(iter);
+
+                listener.onTraversalComplete(System.nanoTime() - lookupStartTime, TimeUnit.NANOSECONDS);
+
+                return new LongHeapPostingList(postings);
+            }
+            catch (Throwable e)
+            {
+                if (!(e instanceof AbortedOperationException))
+                    logger.error(indexContext.logMessage("Failed to execute term query"), e);
+
+                throw Throwables.cleaned(e);
+            }
+        }
+
+        /**
+         * Build an in-memory heap of row ids from the posting lists of the matching terms.
+         * @return an ordered {@link LongHeap} of row ids
+         */
+        private LongHeap materializeResults(Iterator<Pair<ByteSource,Long>> trieEntries) throws IOException
+        {
+            assert trieEntries.hasNext();
+            var heap = new LongHeap(1);
+            try (var postingsInput = IndexFileUtils.instance.openInput(postingsFile);
+                 var postingsSummaryInput = IndexFileUtils.instance.openInput(postingsFile))
+            {
+                do
+                {
+                    Pair<ByteSource, Long> nextTrieEntry = trieEntries.next();
+                    ByteSource key = nextTrieEntry.left;
+                    long rowId = nextTrieEntry.right;
```

Comment:
This should be called `postingsOffset` as it is the offset into the postings file.